src/sentry_logs.c

```diff
@@ -4,9 +4,237 @@
 #include "sentry_options.h"
 #include "sentry_os.h"
 #include "sentry_scope.h"
+#include "sentry_sync.h"
 
 #include <stdarg.h>
 #include <string.h>
+#include <unistd.h>
+
+// TODO think about this
+#ifdef SENTRY_UNITTEST
+#    define QUEUE_LENGTH 5
+#else
+#    define QUEUE_LENGTH 100
+#endif
+#define FLUSH_TIMER 5
+
+typedef struct {
+    sentry_value_t logs[QUEUE_LENGTH];
+    volatile long index;
+    volatile long adding;
+} log_single_buffer_t;
+
+// TODO look at mutex init for inspiration on how to dynamically init
+//  logs as sentry_value_t list using sentry_value_new_list_with_size(...)
+//  (at runtime since function call is not compile-time available)
+static struct {
+    log_single_buffer_t queue;
+    sentry_bgworker_t *timer_worker;
+    volatile long timer_task_submitted;
+    volatile long timer_running;
+    volatile long flushing;
+    sentry_cond_t request_flush;
+    sentry_cond_t adding_done;
+    sentry_mutex_t adding_lock;
+} g_logs_single_state = { { .index = 0 }, .timer_worker = NULL,
+    .timer_task_submitted = 0, .timer_running = 0, .flushing = 0 };
+
+// Forward declaration for timer task function
+static void timer_task_func(void *task_data, void *worker_state);
+
+static void
+flush_logs_single_queue(void)
+{
+    long already_flushing
+        = sentry__atomic_fetch_and_add(&g_logs_single_state.flushing, 1);
+    if (already_flushing
+        || !sentry__atomic_fetch(&g_logs_single_state.queue.index)) {
+        sentry__atomic_fetch_and_add(&g_logs_single_state.flushing, -1);
```

Comment:
If we are already flushing, shouldn't that variable be set by the thread that is flushing? These two conditions should better be separate if statements, because `flushing` just acts as a reentrancy guard, while a 0 `index` probably should reset `flushing` if it was set, because it would mean the timer triggered the flush.